File: fantasy/backend/scraper.py
"""
MLB data scraper for collecting team statistics.
"""
import requests
from bs4 import BeautifulSoup
import time
import logging
from typing import Dict, List, Optional
from models import db, Team
from config import Config
from datetime import datetime

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class MLBScraper:
    """Scraper for MLB team statistics."""
    
    def __init__(self, season=None):
        self.base_url = "https://bdfed.stitch.mlbinfra.com/bdfed/stats/team"
        self.season = season or datetime.now().year
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        })
    
    def scrape_team_stats(self) -> List[Dict]:
        """
        Scrape team pitching statistics vs lefty and righty batters from MLB API.
        
        Returns:
            List of team statistics dictionaries
        """
        logger.info(f"Starting team statistics scraping for {self.season} season...")
        
        teams_data = []
        
        try:
            # Get overall team pitching stats (includes wins/losses)
            overall_stats = self._fetch_overall_pitching_stats()
            
            # Get vs lefty pitching stats
            vs_lefty_stats = self._fetch_vs_lefty_pitching_stats()
            
            # Get vs righty pitching stats  
            vs_righty_stats = self._fetch_vs_righty_pitching_stats()
            
            # Combine all stats by team
            teams_data = self._combine_team_stats(overall_stats, vs_lefty_stats, vs_righty_stats)
            
            logger.info(f"Successfully scraped statistics for {len(teams_data)} teams")
            return teams_data
            
        except Exception as e:
            logger.error(f"Error scraping team stats: {e}")
            # Fall back to sample data if API fails
            logger.warning("Falling back to sample data due to API error")
            return self._get_sample_team_data()
    
    def _fetch_overall_pitching_stats(self) -> Dict:
        """Fetch overall team pitching stats including wins/losses."""
        params = {
            'stitch_env': 'prod',
            'sportId': '1',
            'gameType': 'R',
            'group': 'pitching',
            'order': 'desc',
            'sortStat': 'era',
            'stats': 'season',
            'season': str(self.season),
            'limit': '30',
            'offset': '0'
        }
        
        response = self.session.get(self.base_url, params=params)
        response.raise_for_status()
        data = response.json()
        
        # Convert to dict keyed by team abbreviation
        team_stats = {}
        for team in data.get('stats', []):
            team_stats[team['teamAbbrev']] = team
            
        logger.info(f"Fetched overall pitching stats for {len(team_stats)} teams")
        return team_stats
    
    def _fetch_vs_lefty_pitching_stats(self) -> Dict:
        """Fetch team pitching stats vs lefty batters."""
        params = {
            'stitch_env': 'prod',
            'sportId': '1',
            'gameType': 'R',
            'group': 'pitching',
            'order': 'desc',
            'sortStat': 'era',
            'stats': 'season',
            'season': str(self.season),
            'limit': '30',
            'offset': '0',
            'sitCodes': 'vl'  # vs lefty
        }
        
        response = self.session.get(self.base_url, params=params)
        response.raise_for_status()
        data = response.json()
        
        # Convert to dict keyed by team abbreviation
        team_stats = {}
        for team in data.get('stats', []):
            team_stats[team['teamAbbrev']] = team
            
        logger.info(f"Fetched vs lefty pitching stats for {len(team_stats)} teams")
        return team_stats
    
    def _fetch_vs_righty_pitching_stats(self) -> Dict:
        """Fetch team pitching stats vs righty batters."""
        params = {
            'stitch_env': 'prod',
            'sportId': '1',
            'gameType': 'R',
            'group': 'pitching',
            'order': 'desc',
            'sortStat': 'era',
            'stats': 'season',
            'season': str(self.season),
            'limit': '30',
            'offset': '0',
            'sitCodes': 'vr'  # vs righty
        }
        
        response = self.session.get(self.base_url, params=params)
        response.raise_for_status()
        data = response.json()
        
        # Convert to dict keyed by team abbreviation
        team_stats = {}
        for team in data.get('stats', []):
            team_stats[team['teamAbbrev']] = team
            
        logger.info(f"Fetched vs righty pitching stats for {len(team_stats)} teams")
        return team_stats
    
    def _combine_team_stats(self, overall_stats: Dict, vs_lefty_stats: Dict, vs_righty_stats: Dict) -> List[Dict]:
        """Combine all team stats into a unified format."""
        teams_data = []
        
        # Get all unique team abbreviations
        all_teams = set(overall_stats.keys()) | set(vs_lefty_stats.keys()) | set(vs_righty_stats.keys())
        
        for team_abbrev in all_teams:
            overall = overall_stats.get(team_abbrev, {})
            vs_lefty = vs_lefty_stats.get(team_abbrev, {})
            vs_righty = vs_righty_stats.get(team_abbrev, {})
            
            # Skip if we don't have overall stats (most important)
            if not overall:
                logger.warning(f"No overall stats found for team {team_abbrev}, skipping")
                continue
            
            team_data = {
                'abbreviation': team_abbrev,
                'name': overall.get('teamName', f'Team {team_abbrev}'),
                
                # Vs lefty stats
                'vs_lefty_era': float(vs_lefty.get('era', 0.0)),
                'vs_lefty_whip': float(vs_lefty.get('whip', 0.0)),
                'vs_lefty_k_per_9': self._calculate_per_9(vs_lefty.get('strikeOuts', 0), vs_lefty.get('inningsPitched', '0')),
                'vs_lefty_bb_per_9': self._calculate_per_9(vs_lefty.get('baseOnBalls', 0), vs_lefty.get('inningsPitched', '0')),
                'vs_lefty_hr_per_9': self._calculate_per_9(vs_lefty.get('homeRuns', 0), vs_lefty.get('inningsPitched', '0')),
                'vs_lefty_hits_per_9': self._calculate_per_9(vs_lefty.get('hits', 0), vs_lefty.get('inningsPitched', '0')),
                'vs_lefty_wins': overall.get('wins', 0),  # Use overall wins for now
                'vs_lefty_losses': overall.get('losses', 0),  # Use overall losses for now
                
                # Vs righty stats
                'vs_righty_era': float(vs_righty.get('era', 0.0)),
                'vs_righty_whip': float(vs_righty.get('whip', 0.0)),
                'vs_righty_k_per_9': self._calculate_per_9(vs_righty.get('strikeOuts', 0), vs_righty.get('inningsPitched', '0')),
                'vs_righty_bb_per_9': self._calculate_per_9(vs_righty.get('baseOnBalls', 0), vs_righty.get('inningsPitched', '0')),
                'vs_righty_hr_per_9': self._calculate_per_9(vs_righty.get('homeRuns', 0), vs_righty.get('inningsPitched', '0')),
                'vs_righty_hits_per_9': self._calculate_per_9(vs_righty.get('hits', 0), vs_righty.get('inningsPitched', '0')),
                'vs_righty_wins': overall.get('wins', 0),  # Use overall wins for now
                'vs_righty_losses': overall.get('losses', 0),  # Use overall losses for now
            }
            
            teams_data.append(team_data)
        
        logger.info(f"Combined stats for {len(teams_data)} teams")
        return teams_data
    
    def _calculate_per_9(self, stat_value: int, innings_pitched_str: str) -> float:
        """Calculate per-9-innings rate."""
        try:
            innings_pitched = float(innings_pitched_str)
            if innings_pitched == 0:
                return 0.0
            return round((stat_value * 9) / innings_pitched, 1)
        except (ValueError, TypeError):
            return 0.0
    
    def _get_sample_team_data(self) -> List[Dict]:
        """
        Generate sample team data for development.
        In production, this would scrape real MLB data.
        
        Returns:
            List of sample team data
        """
        teams_data = []
        
        # Sample data for all MLB teams
        team_names = {
            'COL': 'Colorado Rockies',
            'PIT': 'Pittsburgh Pirates',
            'CWS': 'Chicago White Sox',
            'LAA': 'Los Angeles Angels',
            'CLE': 'Cleveland Guardians',
            'TEX': 'Texas Rangers',
            'DET': 'Detroit Tigers',
            'MIN': 'Minnesota Twins',
            'BAL': 'Baltimore Orioles',
            'KC': 'Kansas City Royals',
            'SF': 'San Francisco Giants',
            'BOS': 'Boston Red Sox',
            'HOU': 'Houston Astros',
            'ATL': 'Atlanta Braves',
            'WSH': 'Washington Nationals',
            'SD': 'San Diego Padres',
            'MIL': 'Milwaukee Brewers',
            'CIN': 'Cincinnati Reds',
            'OAK': 'Oakland Athletics',
            'MIA': 'Miami Marlins',
            'SEA': 'Seattle Mariners',
            'TB': 'Tampa Bay Rays',
            'NYM': 'New York Mets',
            'STL': 'St. Louis Cardinals',
            'PHI': 'Philadelphia Phillies',
            'NYY': 'New York Yankees',
            'AZ': 'Arizona Diamondbacks',
            'LAD': 'Los Angeles Dodgers',
            'TOR': 'Toronto Blue Jays',
            'CHC': 'Chicago Cubs'
        }
        
        # Generate realistic sample data for each team
        import random
        random.seed(42)  # For consistent sample data
        
        for abbr, name in team_names.items():
            # Generate realistic pitching stats
            base_era = random.uniform(3.5, 5.5)
            base_whip = random.uniform(1.15, 1.45)
            base_k_per_9 = random.uniform(7.5, 10.5)
            base_bb_per_9 = random.uniform(2.5, 4.0)
            base_hr_per_9 = random.uniform(0.8, 1.5)
            base_hits_per_9 = random.uniform(7.5, 9.5)
            
            # Generate realistic wins/losses (total should be around 162 games)
            total_wins = random.randint(50, 110)
            total_losses = 162 - total_wins
            
            # Split wins/losses between lefty and righty (roughly proportional)
            lefty_win_pct = random.uniform(0.3, 0.7)  # 30-70% of wins vs lefties
            lefty_wins = int(total_wins * lefty_win_pct)
            righty_wins = total_wins - lefty_wins
            
            lefty_loss_pct = random.uniform(0.3, 0.7)  # 30-70% of losses vs lefties  
            lefty_losses = int(total_losses * lefty_loss_pct)
            righty_losses = total_losses - lefty_losses
            
            # Lefty vs righty variations (lefties typically perform slightly better vs righties)
            lefty_modifier = random.uniform(0.95, 1.05)
            righty_modifier = random.uniform(0.95, 1.05)
            
            team_data = {
                'abbreviation': abbr,
                'name': name,
                'vs_lefty_era': round(base_era * lefty_modifier, 2),
                'vs_lefty_whip': round(base_whip * lefty_modifier, 3),
                'vs_lefty_k_per_9': round(base_k_per_9 * lefty_modifier, 1),
                'vs_lefty_bb_per_9': round(base_bb_per_9 * lefty_modifier, 1),
                'vs_lefty_hr_per_9': round(base_hr_per_9 * lefty_modifier, 2),
                'vs_lefty_hits_per_9': round(base_hits_per_9 * lefty_modifier, 1),
                'vs_lefty_wins': lefty_wins,
                'vs_lefty_losses': lefty_losses,
                'vs_righty_era': round(base_era * righty_modifier, 2),
                'vs_righty_whip': round(base_whip * righty_modifier, 3),
                'vs_righty_k_per_9': round(base_k_per_9 * righty_modifier, 1),
                'vs_righty_bb_per_9': round(base_bb_per_9 * righty_modifier, 1),
                'vs_righty_hr_per_9': round(base_hr_per_9 * righty_modifier, 2),
                'vs_righty_hits_per_9': round(base_hits_per_9 * righty_modifier, 1),
                'vs_righty_wins': righty_wins,
                'vs_righty_losses': righty_losses,
            }
            
            teams_data.append(team_data)
        
        return teams_data
    
    def save_team_data(self, teams_data: List[Dict]) -> None:
        """
        Save team data to the database.
        
        Args:
            teams_data: List of team data dictionaries
        """
        logger.info("Saving team data to database...")
        
        for team_data in teams_data:
            # Check if team exists
            team = Team.query.filter_by(abbreviation=team_data['abbreviation']).first()
            
            if team:
                # Update existing team
                for key, value in team_data.items():
                    if hasattr(team, key):
                        setattr(team, key, value)
                logger.info(f"Updated team: {team_data['abbreviation']}")
            else:
                # Create new team
                team = Team(**team_data)
                db.session.add(team)
                logger.info(f"Created team: {team_data['abbreviation']}")
        
        try:
            db.session.commit()
            logger.info("Successfully saved all team data")
        except Exception as e:
            db.session.rollback()
            logger.error(f"Error saving team data: {e}")
            raise
    
    def run_full_scrape(self) -> None:
        """Run the complete scraping process."""
        try:
            teams_data = self.scrape_team_stats()
            self.save_team_data(teams_data)
            logger.info("Full scraping process completed successfully")
        except Exception as e:
            logger.error(f"Error during scraping process: {e}")
            raise


def run_scraper():
    """Standalone function to run the scraper."""
    from app import create_app
    
    app = create_app()
    with app.app_context():
        scraper = MLBScraper()
        scraper.run_full_scrape()


if __name__ == '__main__':
    run_scraper()

File: old/fantasy_react_only/server/scraper.js
const axios = require('axios');
const sqlite3 = require('sqlite3').verbose();
const path = require('path');

class BaseballScraper {
  constructor() {
    this.dbPath = path.join(__dirname, 'fantasy_baseball.db');
    this.db = new sqlite3.Database(this.dbPath);
    this.baseUrl = 'https://www.mlb.com';
  }

  // Scrape player stats from MLB.com or similar source
  async scrapePlayerStats() {
    try {
      console.log('Starting player stats scraping...');
      
      // This is a simplified example - in reality you'd scrape from MLB.com or use their API
      // For now, we'll add some sample data to demonstrate the structure
      const samplePlayers = [
        {
          player_name: 'Mike Trout',
          team: 'LAA',
          position: 'OF',
          handedness: 'Righty',
          hits: 140,
          runs: 85,
          rbi: 95,
          home_runs: 35,
          stolen_bases: 8,
          walks: 90,
          strikeouts: 120,
          games_played: 140,
          at_bats: 520
        },
        {
          player_name: 'Ronald Acuna Jr.',
          team: 'ATL',
          position: 'OF',
          handedness: 'Righty',
          hits: 160,
          runs: 110,
          rbi: 85,
          home_runs: 40,
          stolen_bases: 35,
          walks: 75,
          strikeouts: 140,
          games_played: 150,
          at_bats: 580
        },
        {
          player_name: 'Juan Soto',
          team: 'SD',
          position: 'OF',
          handedness: 'Lefty',
          hits: 145,
          runs: 95,
          rbi: 100,
          home_runs: 30,
          stolen_bases: 12,
          walks: 110,
          strikeouts: 100,
          games_played: 145,
          at_bats: 500
        },
        {
          player_name: 'Mookie Betts',
          team: 'LAD',
          position: 'OF',
          handedness: 'Righty',
          hits: 155,
          runs: 105,
          rbi: 90,
          home_runs: 32,
          stolen_bases: 18,
          walks: 85,
          strikeouts: 110,
          games_played: 148,
          at_bats: 540
        },
        {
          player_name: 'Vladimir Guerrero Jr.',
          team: 'TOR',
          position: '1B',
          handedness: 'Righty',
          hits: 150,
          runs: 80,
          rbi: 110,
          home_runs: 38,
          stolen_bases: 2,
          walks: 70,
          strikeouts: 95,
          games_played: 145,
          at_bats: 550
        }
      ];

      for (const player of samplePlayers) {
        await this.savePlayerStats(player);
      }

      console.log('Player stats scraping completed!');
    } catch (error) {
      console.error('Error scraping player stats:', error);
    }
  }

  // Scrape team pitching stats
  async scrapeTeamStats() {
    try {
      console.log('Starting team stats scraping...');
      
      // Sample team pitching data
      const sampleTeams = [
        {
          team_name: 'Los Angeles Angels',
          vs_lefty_era: 4.25,
          vs_righty_era: 4.10,
          vs_lefty_whip: 1.35,
          vs_righty_whip: 1.28,
          runs_allowed_per_game: 4.8,
          hits_allowed_per_game: 8.5,
          home_runs_allowed_per_game: 1.2,
          walks_allowed_per_game: 3.2,
          strikeouts_per_game: 8.8
        },
        {
          team_name: 'Atlanta Braves',
          vs_lefty_era: 3.85,
          vs_righty_era: 3.70,
          vs_lefty_whip: 1.25,
          vs_righty_whip: 1.20,
          runs_allowed_per_game: 4.2,
          hits_allowed_per_game: 7.8,
          home_runs_allowed_per_game: 1.0,
          walks_allowed_per_game: 2.8,
          strikeouts_per_game: 9.5
        },
        {
          team_name: 'San Diego Padres',
          vs_lefty_era: 4.05,
          vs_righty_era: 3.95,
          vs_lefty_whip: 1.30,
          vs_righty_whip: 1.25,
          runs_allowed_per_game: 4.5,
          hits_allowed_per_game: 8.2,
          home_runs_allowed_per_game: 1.1,
          walks_allowed_per_game: 3.0,
          strikeouts_per_game: 9.2
        },
        {
          team_name: 'Los Angeles Dodgers',
          vs_lefty_era: 3.60,
          vs_righty_era: 3.45,
          vs_lefty_whip: 1.18,
          vs_righty_whip: 1.15,
          runs_allowed_per_game: 3.8,
          hits_allowed_per_game: 7.2,
          home_runs_allowed_per_game: 0.9,
          walks_allowed_per_game: 2.5,
          strikeouts_per_game: 10.1
        },
        {
          team_name: 'Toronto Blue Jays',
          vs_lefty_era: 4.15,
          vs_righty_era: 4.00,
          vs_lefty_whip: 1.32,
          vs_righty_whip: 1.27,
          runs_allowed_per_game: 4.6,
          hits_allowed_per_game: 8.3,
          home_runs_allowed_per_game: 1.3,
          walks_allowed_per_game: 3.1,
          strikeouts_per_game: 8.9
        }
      ];

      for (const team of sampleTeams) {
        await this.saveTeamStats(team);
      }

      console.log('Team stats scraping completed!');
    } catch (error) {
      console.error('Error scraping team stats:', error);
    }
  }

  // Save player stats to database
  savePlayerStats(playerData) {
    return new Promise((resolve, reject) => {
      const sql = `INSERT OR REPLACE INTO player_stats 
        (player_name, team, position, handedness, hits, runs, rbi, home_runs, 
         stolen_bases, walks, strikeouts, games_played, at_bats, updated_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)`;

      this.db.run(sql, [
        playerData.player_name,
        playerData.team,
        playerData.position,
        playerData.handedness,
        playerData.hits,
        playerData.runs,
        playerData.rbi,
        playerData.home_runs,
        playerData.stolen_bases,
        playerData.walks,
        playerData.strikeouts,
        playerData.games_played,
        playerData.at_bats
      ], function(err) {
        if (err) {
          console.error('Error saving player stats:', err);
          reject(err);
        } else {
          console.log(`Saved stats for ${playerData.player_name}`);
          resolve(this.lastID);
        }
      });
    });
  }

  // Save team stats to database
  saveTeamStats(teamData) {
    return new Promise((resolve, reject) => {
      const sql = `INSERT OR REPLACE INTO team_stats 
        (team_name, vs_lefty_era, vs_righty_era, vs_lefty_whip, vs_righty_whip,
         runs_allowed_per_game, hits_allowed_per_game, home_runs_allowed_per_game,
         walks_allowed_per_game, strikeouts_per_game, updated_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)`;

      this.db.run(sql, [
        teamData.team_name,
        teamData.vs_lefty_era,
        teamData.vs_righty_era,
        teamData.vs_lefty_whip,
        teamData.vs_righty_whip,
        teamData.runs_allowed_per_game,
        teamData.hits_allowed_per_game,
        teamData.home_runs_allowed_per_game,
        teamData.walks_allowed_per_game,
        teamData.strikeouts_per_game
      ], function(err) {
        if (err) {
          console.error('Error saving team stats:', err);
          reject(err);
        } else {
          console.log(`Saved stats for ${teamData.team_name}`);
          resolve(this.lastID);
        }
      });
    });
  }

  // Run full scraping process
  async runFullScrape() {
    console.log('Starting full scraping process...');
    await this.scrapePlayerStats();
    await this.scrapeTeamStats();
    console.log('Full scraping process completed!');
    this.close();
  }

  // Close database connection
  close() {
    this.db.close((err) => {
      if (err) {
        console.error('Error closing database:', err);
      } else {
        console.log('Database connection closed.');
      }
    });
  }
}

// Run scraper if called directly
if (require.main === module) {
  const scraper = new BaseballScraper();
  scraper.runFullScrape().catch(console.error);
}

module.exports = BaseballScraper;

